### PR TITLE
gh-101408: PyObject_GC_Resize should calculate preheader size.

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-101408

### CODE REVIEW

File: Modules/gcmodule.c [hunk 1]
```diff
@@ -2348,13 +2348,14 @@ PyVarObject *
...
+    const size_t presize = _PyType_PreHeaderSize(((PyObject *)op)->ob_type);
     _PyObject_ASSERT((PyObject *)op, !_PyObject_GC_IS_TRACKED(op));
-    if (basicsize > (size_t)PY_SSIZE_T_MAX - sizeof(PyGC_Head)) {
+    if (basicsize > (size_t)PY_SSIZE_T_MAX - presize) {
         return (PyVarObject *)PyErr_NoMemory();
     }
 
     PyGC_Head *g = AS_GC(op);
-    g = (PyGC_Head *)PyObject_Realloc(g,  sizeof(PyGC_Head) + basicsize);
+    g = (PyGC_Head *)PyObject_Realloc(g,  presize + basicsize);
```
Comment (colesbury) [inline Modules/gcmodule.c]: These lines (2357-2361) will not be correct if `presize` includes a managed dict. The layout would look like:

```
[ weak ptr ] [ managed dict ] [ _gc_next ] [ _gc_prev ] [PyObject ....]
^alloc                        ^ g                       ^ op
```

The bugs (when including manage dict/weak ptr) are:
1. You can't call PyObject_Realloc on `g`, it needs to be called on what I labelled as `alloc` in my diagram above.
2. The returned value from `PyObject_Realloc` will be the beginning of the allocation (not necessarily the PyGC_Head)

Comment (corona10) [reply Modules/gcmodule.c]: Ah thanks for the pointing out. I will take a look.

**Review by colesbury — COMMENTED** [review]
This LGTM!

Minor: I don't think this needs backport to 3.10 or 3.11. I think the preheader feature was introduced in 3.12.

File: Modules/gcmodule.c (hunk 1 above)
Comment (corona10) [reply Modules/gcmodule.c]: @colesbury 
I've updated the PR. Thanks for pointing it out; this is my first patch related to GC.
I wish that my changes handle the object header properly.

**Comment by corona10** [comment]
The refleak doesn't related to this PR, I am finding out the which PR cause the leak.

**Comment by pablogsal** [comment]
Thanks a lot for tacking this @corona10 and thanks for the catch @colesbury!

I would be more comfortable if we could have a test in `testcapi` that uses this API, among other things so we don't regress in the future if we keep changing things.

**Comment by corona10** [comment]
> I would be more comfortable if we could have a test in testcapi that uses this API, among other things so we don't regress in the future if we keep changing things.

Okay, I will finalize the PR by end of this weekend. :)

**Comment by corona10** [comment]
@pablogsal 

Hi Pablo, I succeeded in adding the test code for the normal PyObject_GC_Resize case.
But looks hard to create the test code for `Py_TPFLAGS_PREHEADER`. 
Do you have any recommended way to for this case?

`````C

static PyObject *
test_pyobject_gc(PyObject *self, PyObject *Py_UNUSED(ignored))
{
    PyTupleObject *obj = PyObject_GC_NewVar(PyTupleObject, &PyTuple_Type, 3);
    obj->ob_item[0] = Py_NewRef(Py_None);
    obj->ob_item[1] = Py_NewRef(Py_None);
    obj->ob_item[2] = Py_NewRef(Py_None);

    if (obj == NULL) {
        goto alloc_failed;
    }
    if (Py_SIZE(obj) != 3) {
        Py_DECREF(obj);
        PyErr_SetString(PyExc_RuntimeError, "Invalid Py_SIZE(obj)");
        return NULL;
    }

    PyTupleObject *new_obj = PyObject_GC_Resize(PyTupleObject, obj, 6);
    if (new_obj == NULL) {
        Py_DECREF(obj);
        goto alloc_failed;
    }

    new_obj->ob_item[3] = Py_NewRef(Py_None);
    new_obj->ob_item[4] = Py_NewRef(Py_None);
    new_obj->ob_item[5] = Py_NewRef(Py_None);
    if (Py_SIZE(new_obj) != 6) {
        Py_DECREF(new_obj);
        PyErr_SetString(PyExc_RuntimeError, "Invalid Py_SIZE(new_obj)");
        return NULL;
    }

    Py_DECREF(new_obj);
    Py_RETURN_NONE;

alloc_failed:
    PyErr_NoMemory();
    return NULL;
}
`````

**Comment by corona10** [comment]
@pablogsal I think that this PR should be merged before 3.12 is officially released, I will let the issue as opened for test code and merging this PR.

**Comment by vstinner** [comment]
No one wrote code since April, maybe it's time to abandon the idea of having tests and just close the issue?

**Pull request merged**